create script to install latest powershell from repositories

DarwinJS opened the pull request:
I am an intermediate-noob at bash so I would love to have some testing of the new installer before doing a pull request.

The goals and overall idea is described in #3547 

## Features (and fixes of download.sh):
* Bare minimum requirements to get going (wget & sed) (attempts to install curl if not present)
* defaults to only install PowerShell core (using repos where they are available)
* For non-repo based installs - detects latest version from git tags
* Optional -includeide switch installs VS Code and PowerShell Extension (using repos where they are available)
* Fix/Cleanup: Installs on OpenSUSE using download from regular github releases page (until repo is established)
* Removed restriction to OpenSUSE 42.1, should install 42.1 and later.
* Removed restriction to CentOS only - should install on Redhat as well.
* Uses brew repository for OSX.

## I have it working and tested under:
* Ubuntu Desktop 16.04
* CentOS 7.2 with GNOME Desktop
* OpenSUSE 42.2 with KDE Desktop

## Known Issues (and Help Needed):
* The -appimage switch just calls the pre-existing appimage.sh which did not work on any of the above test cases
* The OSX install was pieced together from various sources and is untested.

Here is the readme: https://github.com/DarwinJS/PowerShell/blob/feature/install-powershell.sh/tools/install-powershell.readme.md

## Here are the commands to kick off the install (also covered in readme):

### Only Install PowerShell Core:
```bash
bash <(wget -O - https://raw.githubusercontent.com/DarwinJS/PowerShell/feature/install-powershell.sh/tools/install-powershell.sh)
```
###  Install PowerShell Core with IDE:
```bash
bash <(wget -O - https://raw.githubusercontent.com/DarwinJS/PowerShell/feature/install-powershell.sh/tools/install-powershell.sh) -includeide
```
###  Install PowerShell Core with IDE and do tests that require a human to interact with the install process:
```bash
bash <(wget -O - https://raw.githubusercontent.com/DarwinJS/PowerShell/feature/install-powershell.sh/tools/install-powershell.sh) -includeide -interactivetesting
```

msftclas commented:
@DarwinJS,
Thanks for having already signed the Contribution License Agreement. Your agreement was validated by Microsoft. We will now review your pull request.
_Thanks,
Microsoft Pull Request Bot_

JamesWTruher commented on tools/install-powershell.sh:
```diff
@@ -0,0 +1,137 @@
+
+lowercase(){
+    echo "$1" | sed "y/ABCDEFGHIJKLMNOPQRSTUVWXYZ/abcdefghijklmnopqrstuvwxyz/"
+}
```
I think this can be more easily done as 
`echo "$*" | tr [A-Z] [a-z]`

DarwinJS replied:
@JamesWTruher - this came over from download.sh.  Do you know if "tr" universally available on al platform targets?

DarwinJS replied:
Ok did some research and it seems like anything in the posix command set should be considered universal and "tr" is on the list: http://pubs.opengroup.org/onlinepubs/9699919799/utilities/contents.html

JamesWTruher commented on tools/install-powershell.sh:
```diff
@@ -0,0 +1,137 @@
+    OS=osx
+    DistroBasedOn=osx
+else
+    OS=`uname`
```
could the rest of these comparisons all be lower case comparisons? is there a need to rerun uname?

JamesWTruher commented on tools/install-powershell.sh:
```diff
@@ -0,0 +1,137 @@
+    elif [ "${OS}" == "Linux" ] ; then
+        if [ -f /etc/redhat-release ] ; then
+            DistroBasedOn='redhat'
+            DIST=`cat /etc/redhat-release |sed s/\ release.*//`
```
I think that `cat` is unneeded. `sed` and `grep` all take a file so this could just be 
`DIST=$(sed /s\ release.*// /etc/redhat-release)`

JamesWTruher commented on tools/installpsh-debian.sh:
```diff
@@ -0,0 +1,167 @@
+
+#Verify The Installer Choice (for direct runs of this script)
+lowercase(){
+    echo "$1" | sed "y/ABCDEFGHIJKLMNOPQRSTUVWXYZ/abcdefghijklmnopqrstuvwxyz/"
```
>echo "$1" | sed "y/ABCDEFGHIJKLMNOPQRSTUVWXYZ/abcdefghijklmnopqrstuvwxyz/" [](start = 4, length = 74)

`echo "$1" | tr [A-Z] [a-z]`

JamesWTruher commented on tools/installpsh-debian.sh:
```diff
@@ -0,0 +1,167 @@
+
+echo
+echo "*** Installing PowerShell Core for $DistroBasedOn..."
+if ! hash curl 2>/dev/null; then
```
>hash [](start = 5, length = 4)

i know this works, but would `which` be a better choice?

JamesWTruher replied:
>curl [](start = 10, length = 4)

should you also check for `wget`?

DarwinJS replied:
Not sure - is "which" available on all platform targets including OSX?  I will tend to favor what is working over making changes to something that in theory is a better method.

DarwinJS replied:
In my googling on it seemed that wget is pretty universal - this was also born out by the original download.sh.  But I like the idea, because if someone replicates the repo or copies the script set manually - that is a scenario where I can't assume wget was used to pull the script in the first place.

JamesWTruher commented on tools/installpsh-debian.sh:
```diff
@@ -0,0 +1,167 @@
+    echo "curl not found, installing..."
+    $SUDO apt-get install -y curl
+fi
+release=`curl https://api.github.com/repos/powershell/powershell/releases/latest | sed '/tag_name/!d' | sed s/\"tag_name\"://g | sed s/\"//g | sed s/v//g | sed s/,//g | sed s/\ //g`
```
>curl https://api.github.com/repos/powershell/powershell/releases/latest | sed '/tag_name/!d' | sed s/\"tag_name\"://g | sed s/\"//g | sed s/v//g | sed s/,//g | sed s/\ //g` [](start = 9, length = 172)

there's all sorts of ways to do this. here's an alternative:
`curl -q [https://api.github.com/repos/powershell/powershell/releases/latest](https://api.github.com/repos/powershell/powershell/releases/latest) 2>/dev/null | awk ' /tag_name/ { print $2 }' | sed "s/[\",]//g" `

DarwinJS replied:
I'm sure there are a lot of ways any given thing could be done - really looking for the smallest possible toolset dependencies to get the practical result.

JamesWTruher commented on tools/install-powershell.sh:
```diff
@@ -0,0 +1,137 @@
+KERNEL=`uname -r`
+MACH=`uname -m`
+
+if [ "{$OS}" == "windowsnt" ]; then
```
>"{ [](start = 5, length = 2)

this should be `"${OS}"` I think

DarwinJS replied:
Thanks!

JamesWTruher commented on tools/install-powershell.sh:
```diff
@@ -0,0 +1,137 @@
+if [ "{$OS}" == "windowsnt" ]; then
+    OS=windows
+    DistroBasedOn=windows
+elif [ "{$OS}" == "darwin" ]; then
```
>{$OS} [](start = 8, length = 5)

`"${OS}"`

JamesWTruher commented on tools/installpsh-osx.sh:
```diff
@@ -0,0 +1,206 @@
+
+#Verify The Installer Choice (for direct runs of this script)
+lowercase(){
+    echo "$1" | sed "y/ABCDEFGHIJKLMNOPQRSTUVWXYZ/abcdefghijklmnopqrstuvwxyz/"
```
>sed [](start = 16, length = 3)

`tr`

JamesWTruher commented on tools/installpsh-osx.sh:
```diff
@@ -0,0 +1,206 @@
+lowercase(){
+    echo "$1" | sed "y/ABCDEFGHIJKLMNOPQRSTUVWXYZ/abcdefghijklmnopqrstuvwxyz/"
+}
+if [ "{$OS}" == "windowsnt" ]; then
```
"${OS}"

JamesWTruher commented on tools/installpsh-osx.sh:
```diff
@@ -0,0 +1,206 @@
+if [ "{$OS}" == "windowsnt" ]; then
+    OS=windows
+    DistroBasedOn=windows
+elif [ "{$OS}" == "darwin" ]; then
```
>$OS [](start = 9, length = 3)

"${OS}"

JamesWTruher commented on tools/installpsh-suse.sh:
```diff
@@ -0,0 +1,190 @@
+gitreposcriptroot="https://raw.githubusercontent.com/$gitreposubpath/tools"
+thisinstallerdistro=suse
+repobased=true
+gitscriptname="installpsh-redhat.psh"
```
>redhat [](start = 26, length = 6)

`suse`?

DarwinJS replied:
Good catch - thanks!

DarwinJS commented on tools/install-powershell.readme.md:
```diff
@@ -0,0 +1,56 @@
+
+## Features of install-powershell.sh
```
@TravisEz13 - I'm not sure what you mean.  "##" is level two heading.

I also don't know what you mean by "Can you please enable these files in this test..."  - how would I do this?

Please point me to a markdown validator where I can reproduce the syntax errors or the failed test report.

Thanks,
D.

DarwinJS replied:
@TravisEz13 - I have tried to run the markdown syntax test locally.  I have installed node-js and gulp as per the instructions in the script you linked, but it does not recognize 'test-mdsyntax' and I don't find it in the repo.  I'm on Windows 10.

DarwinJS commented:
@TravisEz13 - I read the contributing information for pull requests and it says "clean up your commit history"

I did that by rebasing - but now I'm wondering if the idea is to keep at least one new commit for each update to the PR?

Let me know whether I should be squashing into my originally submited commit or something else.

iSazonov commented:
@DarwinJS We must preserve the history, we should avoid rebaseing for _already published_ commits. For new commits we should do one commit for one logical work piece: putting 10 commits with removing extra lines is bad.

DarwinJS commented:
@TravisEz13 - I added to spelling exceptions to the repo root's .spelling - but they seem to be ignored when checking my .md in the tools subfolder.

Should I create a .spelling in the tools subfolder - or is there something that has to be adjusted about the calling to the spelling checker?

iSazonov commented:
@DarwinJS Maybe you can do spell checking on WSL.

DarwinJS commented:
@iSazonov - I neglected to mention I ran spell checking locally with the exact same code as used by CI and generated two spelling exceptions which it automatically added to .spelling in the repo root.  However, when travis runs the build it does not seem to pickup on that .spelling file when it is processing the tools folder - so just trying to figure our the correct next step.

DarwinJS commented:
@joeyaiello - this PR is being blocked by the "spelling errors" on 'sed' and '-includeide' in the md help document I put together.

I have these exceptions in the main .spelling exception file and copied the .spelling file to the tools subfolder - still won't make it through.

What else might I do to pass the checks?

The command "powershell -File tools/travis.ps1" exited with 0.
1.83s$ if [[ "$TRAVIS_OS_NAME" == "linux" ]]; then mdspell '**/*.md' '!powershell/**/*.md' --ignore-numbers --ignore-acronyms --report; fi
    tools/install-powershell.readme.md
        6 | werShell IDE) using optional -includeide switch 
       18 | * sed

iSazonov commented:
@DarwinJS You added "includeide" - try "-includeide".
For "sed": 1. try add Newline after "sed" 2. Try "* sed"

DarwinJS commented:
@iSazonov - I have tried you solutions in both .spelling files.  There already was a newline after sed and the original additions to the file was done by interactively running mdspell.

iSazonov commented:
We can wrap "-includeide" with apostrophes `` in line 6 (and remove from `.spelling`)
But I dont understand that problem with "sed". Maybe use apostrophes too.

DarwinJS commented:
mdspell runs clean on that file on my local system (windows 10):

![image](https://cloud.githubusercontent.com/assets/11597190/25463787/a088390a-2ac6-11e7-857d-45f5fdeab9bd.png)

Is the version running in CI the latest?

I am out of ideas :(

iSazonov commented:
@DarwinJS It seems a mdspell issue. Did you try to use apostrophes?

DarwinJS committed:
trying apostrophes
rename md to have only one period
fix md spelling mistakes
Fixes to markdown to follow strict syntax
create readme

DarwinJS commented:
@TravisEz13 and @iSazonov  - Phew - Finally!

DarwinJS committed:
install-powershell.sh
updating for pull request comments
OSX Install script updated to use repositories

Using TR, use curl rather than wget since we ensured it is on the system
Update comments in install-powershell.sh
install-powershell.sh

DarwinJS commented:
I no longer understand what is holding up this PR - can someone please enlighten me?  Thanks!

iSazonov commented:
Please resolve file conflicting.

DarwinJS commented:
@TravisEz13 - I am interested in being listed as a contributor in this repo - will that still be the case with the actions you've taken?

DarwinJS commented:
Thanks for your explanation and for your help @TravisEz13  !

TravisEz13 committed:
Get macOS working and fix misc issue

DarwinJS commented:
@TravisEz13 - thanks for all the fixups - exactly what I was asking for in one of the previous conference calls!

DarwinJS commented:
@TravisEz13 - maybe this should just call install-powershell.sh now?

DarwinJS commented:
@TravisEz13 - apologies got this confused with one for download.sh
